File: cute-bedwars-database-default-impl-sqlite/src/main/java/io/github/gdrfgdrf/cutebedwars/database/impl/sqlite/service/base/BetterServiceImpl.java
package io.github.gdrfgdrf.cutebedwars.database.impl.sqlite.service.base;

import com.baomidou.mybatisplus.core.mapper.BaseMapper;
import com.github.dreamyoung.mprelation.ServiceImpl;
import io.github.gdrfgdrf.cutebedwars.abstracts.utils.IClasses;
import io.github.gdrfgdrf.cutebedwars.database.impl.sqlite.common.Mappers;

/**
 * @author gdrfgdrf
 */
public class BetterServiceImpl<M extends BaseMapper<T>, T> extends ServiceImpl<M, T> {
    private M mapperInstance;

    @SuppressWarnings("unchecked")
    public M mapper() {
        if (mapperInstance != null) {
            return mapperInstance;
        }

        Class<?> mType = IClasses.Companion.instance().getClassParameter(this, 0);
        mapperInstance = Mappers.INSTANCE.getOrCreateMapper((Class<? extends BaseMapper<?>>) mType);
        return mapperInstance;
    }

    public int insert(T t) {
        if (mapper() == null) {
            return -1;
        }
        return mapper().insert(t);
    }

}

File: cute-bedwars-database-default-impl-mysql/src/main/java/io/github/gdrfgdrf/cutebedwars/database/impl/mysql/service/base/BetterServiceImpl.kt
package io.github.gdrfgdrf.cutebedwars.database.impl.mysql.service.base

import com.baomidou.mybatisplus.core.mapper.BaseMapper
import com.github.dreamyoung.mprelation.ServiceImpl
import io.github.gdrfgdrf.cutebedwars.abstracts.utils.logInfo
import io.github.gdrfgdrf.cutebedwars.database.impl.mysql.common.Mappers

open class BetterServiceImpl<M : BaseMapper<T>, T>: ServiceImpl<M, T>() {
    var mapper: M? = null
        get() {
            if (field == null) {
                field = Mappers.getOrCreateMapper(mapperClass)
            }
            return field
        }

    fun insert(t: T): Int {
        val result = mapper?.insert(t) ?: return -1
        return result
    }

    init {
        "Service ${this::class.java.name} is created".logInfo()
    }
}